File: cc/040.cc
// Project Euler Problem 040
//
// Finding the nth digit of the fractional part of the irrational
// number.

#include <cstdio>
#include <cmath>

int nth_digit(int n, int starts[]) {
  int zn = n-1;
  int i = 0;
  while (zn >= starts[i+1]) {
    i++;
  }
  int count = zn - starts[i];
  int number = count / (i+1) + pow(10, i);
  int div = pow(10, i - (count % (i+1)));
  return (number / div) % 10;
}

int main(int argc, const char* argv[]) {
  int starts[7] = { 0 };
  int t1 = 1;
  for (int i=0; i<6; i++) {
    int count = 9 * t1;
    starts[i+1] = starts[i] + (i+1) * count;
    t1 *= 10;
  }

  int prod = 1;
  int d = 10;
  for (int i = 0; i<6; i++) {
    prod *= nth_digit(d, starts);
    d *= 10;
  }
  printf("%d\n", prod);
}

File: go/problems/040.go
// Project Euler Problem 040
//
// Finding the nth digit of the fractional part of the irrational
// number.

package problems

import "fmt"

func Problem040() string {
	c := make(chan int)
	quit := make(chan bool)

	go func() {
		pow10 := 1
		for n := 1; ; n++ {
			if n >= pow10*10 {
				pow10 *= 10
			}
			for nn, pp := n, pow10; pp > 0; {
				digit := nn / pp
				nn -= pp * digit
				pp /= 10
				select {
				case <-quit:
					return
				case c <- digit:
				}
			}
		}
	}()

	prod, pos := 1, 1
	for i := 1; i < 1000001; i++ {
		digit := <-c
		if i == pos {
			prod *= digit
			pos *= 10
		}
	}
	quit <- true
	return fmt.Sprintf("%d", prod)
}

func init() {
	Functions[40] = Problem040
}
